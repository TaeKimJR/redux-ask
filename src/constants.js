const REQUEST_REDUCER_IDENTIFIER = 'REQUEST_REDUCER';

const NOT_STARTED = 'NOT_STARTED';
const PENDING = 'PENDING';
const SUCCESS = 'SUCCESS';
const FAILURE = 'FAILURE';

const status = {
  NOT_STARTED,
  PENDING,
  SUCCESS,
  FAILURE,
};

export {
  REQUEST_REDUCER_IDENTIFIER,
  status,
};
