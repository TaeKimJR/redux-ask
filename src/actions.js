import { createAction } from 'redux-actions';
import { createInfiniteAction } from 'infinite-reducer';
import { REQUEST_REDUCER_IDENTIFIER } from './constants';
import * as types from './types';

const setInitialUnwrapped = createAction(types.SET_INITIAL);
const setPendingUnwrapped = createAction(types.SET_PENDING);
const setSuccessUnwrapped = createAction(types.SET_SUCCESS);
const setFailureUnwrapped = createAction(types.SET_FAILURE);

export const unwrapped = {
  setInitial: setInitialUnwrapped,
  setPending: setPendingUnwrapped,
  setSuccess: setSuccessUnwrapped,
  setFailure: setFailureUnwrapped,
};

const setInitial = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setInitialUnwrapped);
const setPending = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setPendingUnwrapped);
const setSuccess = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setSuccessUnwrapped);
const setFailure = createInfiniteAction(REQUEST_REDUCER_IDENTIFIER, setFailureUnwrapped);

export {
  setInitial,
  setPending,
  setSuccess,
  setFailure,
};
