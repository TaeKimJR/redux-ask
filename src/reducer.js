import * as R from 'ramda';
import { handleActions } from 'redux-actions';
import { combineReducers } from 'redux';
import { createInfiniteReducer } from 'infinite-reducer';
import { REQUEST_REDUCER_IDENTIFIER, status } from './constants';
import * as types from './types';

// Used for each individual requests
const requestInitialState = {
  status: status.NOT_STARTED,
  response: null,
  error: null,
};
export const requestReducer = handleActions(
  {
    [types.SET_INITIAL]: R.always(requestInitialState),
    [types.SET_PENDING]: state => R.assoc('status', status.PENDING, state),
    [types.SET_SUCCESS]: (state, { payload }) => R.compose(
      R.assoc('response', payload),
      R.assoc('status', status.SUCCESS),
    )(state),
    [types.SET_FAILURE]: (state, { payload }) => R.compose(
      R.assoc('error', payload),
      R.assoc('status', status.FAILURE),
    )(state),
  },
  requestInitialState,
);

// Used for all requests, sharing the requestReducer
const requestReducers = createInfiniteReducer(REQUEST_REDUCER_IDENTIFIER, requestReducer);


// Used for configuring the request api
const configInitialState = {
  options: {},
  onUnauthenticated: null,
  onSuccess: null,
  onFailure: null,
};

export const configReducer = handleActions(
  {
    [types.SET_CONFIG]: (state, { payload }) => R.merge(configInitialState, payload),
  },
  configInitialState,
);

// Used to create the reducer (allowing for configuration overrides passed in)
const reducer = combineReducers({
  config: configReducer,
  requests: requestReducers,
});

export default reducer;

