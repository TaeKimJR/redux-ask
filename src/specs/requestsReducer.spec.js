import { requestReducer } from '../reducer';
import { unwrapped as actions } from '../actions';
import { status } from '../constants';

describe('[Requests] Request Reducer', () => {
  describe('SET_INITIAL', () => {
    test('should set the state back to initial state', () => {
      const state = {
        status: status.SUCCESS,
        response: { foo: 'bar' },
        error: null,
      };

      const reduced = requestReducer(state, actions.setInitial());

      expect(reduced).toMatchSnapshot();
    });
  });

  describe('SET_PENDING', () => {
    test('should set the state to pending', () => {
      const reduced = requestReducer(null, actions.setPending());

      expect(reduced).toMatchSnapshot();
    });
  });

  describe('SET_SUCCESS', () => {
    test('should set the status and response object given', () => {
      const response = { foo: 'bar' };

      const reduced = requestReducer(null, actions.setSuccess(response));

      expect(reduced).toMatchSnapshot();
    });
  });

  describe('SET_FAILURE', () => {
    test('should set the status and error object given', () => {
      const error = { foo: 'bar' };

      const reduced = requestReducer(null, actions.setFailure(error));

      expect(reduced).toMatchSnapshot();
    });
  });
});
