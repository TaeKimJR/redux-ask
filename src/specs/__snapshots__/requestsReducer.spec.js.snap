// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Requests] Request Reducer SET_FAILURE should set the status and error object given 1`] = `
Object {
  "error": Object {
    "foo": "bar",
  },
  "status": "FAILURE",
}
`;

exports[`[Requests] Request Reducer SET_INITIAL should set the state back to initial state 1`] = `
Object {
  "error": null,
  "response": null,
  "status": "NOT_STARTED",
}
`;

exports[`[Requests] Request Reducer SET_PENDING should set the state to pending 1`] = `
Object {
  "status": "PENDING",
}
`;

exports[`[Requests] Request Reducer SET_SUCCESS should set the status and response object given 1`] = `
Object {
  "response": Object {
    "foo": "bar",
  },
  "status": "SUCCESS",
}
`;
